<a href="https://colab.research.google.com/github/matsilvestre/percentual_nulidade_spark/blob/main/percentual_nulidade_atributos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

- Instalando o pyspark


In [ ]:
!pip install pyspark

- Importando as bibliotecas

In [ ]:
from google.colab import drive
from pyspark.sql import SparkSession
from pyspark.sql import functions  as F
from pyspark.sql.types import *

- Configurando Google Drive

In [ ]:
drive.mount('/content/drive')

Mounted at /content/drive


- Criando Sessão Spark

In [ ]:
spark = SparkSession.builder.appName('PERCENTUAL_NULIDADE').getOrCreate()

- Realizando a leitura do arquivo do google drive. Informar o diretório do google drive onde você salvou o arquivo de marketing campaign. Neste exemplo o diretório é: '/content/drive/My Drive/cafe_c_codigo_01/marketing_campaign.csv'

In [ ]:
df = spark.read.csv('/content/drive/My Drive/cafe_c_codigo_01/marketing_campaign.csv', header = True, sep=';')

In [ ]:
df.show(1, vertical = True, truncate = True)

-RECORD 0-------------------------
 ID                  | 5524       
 Year_Birth          | 1957       
 Education           | Graduation 
 Marital_Status      | Single     
 Income              | 58138      
 Kidhome             | 0          
 Teenhome            | 0          
 Dt_Customer         | 2012-09-04 
 Recency             | 58         
 MntWines            | 635        
 MntFruits           | 88         
 MntMeatProducts     | 546        
 MntFishProducts     | 172        
 MntSweetProducts    | 88         
 MntGoldProds        | 88         
 NumDealsPurchases   | 3          
 NumWebPurchases     | 8          
 NumCatalogPurchases | 10         
 NumStorePurchases   | 4          
 NumWebVisitsMonth   | 7          
 AcceptedCmp3        | 0          
 AcceptedCmp4        | 0          
 AcceptedCmp5        | 0          
 AcceptedCmp1        | 0          
 AcceptedCmp2        | 0          
 Complain            | 0          
 Z_CostContact       | 3          
 Z_Revenue          

- Verificando quais atributos tem valores nulos

In [ ]:
df.select(*(F.sum(F.col(c).isNull().cast("int")).alias(c) for c in df.columns)).show(vertical = True)

-RECORD 0------------------
 ID                  | 0   
 Year_Birth          | 0   
 Education           | 0   
 Marital_Status      | 0   
 Income              | 24  
 Kidhome             | 0   
 Teenhome            | 0   
 Dt_Customer         | 0   
 Recency             | 0   
 MntWines            | 0   
 MntFruits           | 0   
 MntMeatProducts     | 0   
 MntFishProducts     | 0   
 MntSweetProducts    | 0   
 MntGoldProds        | 0   
 NumDealsPurchases   | 0   
 NumWebPurchases     | 0   
 NumCatalogPurchases | 0   
 NumStorePurchases   | 0   
 NumWebVisitsMonth   | 0   
 AcceptedCmp3        | 0   
 AcceptedCmp4        | 0   
 AcceptedCmp5        | 0   
 AcceptedCmp1        | 0   
 AcceptedCmp2        | 0   
 Complain            | 0   
 Z_CostContact       | 0   
 Z_Revenue           | 0   
 Response            | 0   



Somente o atributo Income contém valores nulos com total de 24 registros

- Calculo do percentual da nulidade dos campos do resultado da query SQL usando list comprehension

In [ ]:
df.select(*(F.round(((F.sum(F.col(c).isNull().cast("int"))/df.count())*100),2).alias(c) for c in df.columns)).show(vertical = True)

-RECORD 0-------------------
 ID                  | 0.0  
 Year_Birth          | 0.0  
 Education           | 0.0  
 Marital_Status      | 0.0  
 Income              | 1.07 
 Kidhome             | 0.0  
 Teenhome            | 0.0  
 Dt_Customer         | 0.0  
 Recency             | 0.0  
 MntWines            | 0.0  
 MntFruits           | 0.0  
 MntMeatProducts     | 0.0  
 MntFishProducts     | 0.0  
 MntSweetProducts    | 0.0  
 MntGoldProds        | 0.0  
 NumDealsPurchases   | 0.0  
 NumWebPurchases     | 0.0  
 NumCatalogPurchases | 0.0  
 NumStorePurchases   | 0.0  
 NumWebVisitsMonth   | 0.0  
 AcceptedCmp3        | 0.0  
 AcceptedCmp4        | 0.0  
 AcceptedCmp5        | 0.0  
 AcceptedCmp1        | 0.0  
 AcceptedCmp2        | 0.0  
 Complain            | 0.0  
 Z_CostContact       | 0.0  
 Z_Revenue           | 0.0  
 Response            | 0.0  



O atributo Income contém um percentual de nulidade de 1.07%